File: Java/src/Arrays/kadaneAlgo.java
package Arrays;
//https://practice.geeksforgeeks.org/problems/kadanes-algorithm-1587115620/1
//https://leetcode.com/problems/maximum-subarray/
//Given an integer array nums, find the contiguous subarray (containing at least one number)
//which has the largest sum and return its sum. A subarray is a contiguous part of an array.
//Difficulty: Medium
public class kadaneAlgo {
    public static void main(String[] args) {
        int[] arr={-2,1,-3,4,-1,2,1,-5,4};
        System.out.println(maxSubArray(arr));
    }

    //Brute force approach- to check every subarray and find the max of that, o(n^2) or O(n^3)
    //So we use Kadane's algo -> O(n)
    //Concept- We traverse linearly and calculate the sum, whenever the sum gets negative, we discard that part and move on to next part
    static int maxSubArray(int[] nums){
        int maxSum=nums[0];
        int curSum=0;
        for (int i = 0; i < nums.length; i++) {
            curSum=curSum+nums[i];
            if(curSum>maxSum){
                maxSum=curSum;
            }
            if(curSum<0){
                curSum=0;
            }
        }
        return maxSum;
    }
}

File: kadane_algo.cpp
#include<bits/stdc++.h>
using namespace std;

int kadane_algo(int arr[], int n){
        int max_sum = arr[0], sum = 0;
        for(int i = 0; i < n; i++){
            sum = sum + arr[i];
            if(max_sum < sum){
                max_sum = sum;
            }
            if(sum < 0){
                sum = 0;
            }
        }
        return max_sum;
    }
int main()
{
    int arr[100], n, sum=0, i;
    cin>>n;

    for(i = 0; i < n; i++){
        cin >> arr[i];
    }
    int max_sum_final = kadane_algo(arr, n);
    cout << max_sum_final << endl;
    
}